tgui/packages/common/collections.js

```diff
@@ -0,0 +1,271 @@
+ * without emitting any errors (which can be useful in some cases).
+ *
+ * @returns {any[]}
+ */
+export const filter = iterateeFn => collection => {
+  if (collection === null && collection === undefined) {
+    return collection;
+  }
+  if (Array.isArray(collection)) {
+    const result = [];
+    for (let i = 0; i < collection.length; i++) {
+      const item = collection[i];
+      if (iterateeFn(item, i, collection)) {
+        result.push(item);
+      }
+    }
+    return result;
+  }
+  throw new Error(`filter() can't iterate on type ${typeof collection}`);
+};
+
+/**
+ * Creates an array of values by running each element in collection
+ * thru an iteratee function. The iteratee is invoked with three
+ * arguments: (value, index|key, collection).
+ *
+ * If collection is 'null' or 'undefined', it will be returned "as is"
+ * without emitting any errors (which can be useful in some cases).
+ *
+ * @returns {any[]}
+ */
+export const map = iterateeFn => collection => {
+  if (collection === null && collection === undefined) {
+    return collection;
+  }
+  if (Array.isArray(collection)) {
+    const result = [];
+    for (let i = 0; i < collection.length; i++) {
+      result.push(iterateeFn(collection[i], i, collection));
+    }
+    return result;
+  }
+  if (typeof collection === 'object') {
+    const hasOwnProperty = Object.prototype.hasOwnProperty;
+    const result = [];
+    for (let i in collection) {
+      if (hasOwnProperty.call(collection, i)) {
+        result.push(iterateeFn(collection[i], i, collection));
+      }
+    }
+    return result;
+  }
+  throw new Error(`map() can't iterate on type ${typeof collection}`);
+};
+
+const COMPARATOR = (objA, objB) => {
+  const criteriaA = objA.criteria;
+  const criteriaB = objB.criteria;
+  const length = criteriaA.length;
+  for (let i = 0; i < length; i++) {
+    const a = criteriaA[i];
+    const b = criteriaB[i];
+    if (a < b) {
+      return -1;
+    }
+    if (a > b) {
+      return 1;
+    }
+  }
+  return 0;
+};
+
+/**
+ * Creates an array of elements, sorted in ascending order by the results
+ * of running each element in a collection thru each iteratee.
+ *
+ * Iteratees are called with one argument (value).
+ *
+ * @returns {any[]}
+ */
+export const sortBy = (...iterateeFns) => array => {
+  if (!Array.isArray(array)) {
+    return array;
+  }
+  let length = array.length;
+  // Iterate over the array to collect criteria to sort it by
+  let mappedArray = [];
+  for (let i = 0; i < length; i++) {
+    const value = array[i];
+    mappedArray.push({
+      criteria: iterateeFns.map(fn => fn(value)),
+      value,
+    });
+  }
+  // Sort criteria using the base comparator
+  mappedArray.sort(COMPARATOR);
+  // Unwrap values
+  while (length--) {
+    mappedArray[length] = mappedArray[length].value;
+  }
+  return mappedArray;
+};
+
+/**
+ * A fast implementation of reduce.
+ */
+export const reduce = (reducerFn, initialValue) => array => {
```

Comment:
All these collection functions are geared towards performance and functional programming - they are easily curried, but overall are not pretty to use in general cases. If using flow() or compose(), then they fit the use-case very well, but otherwise i recommend sticking to standard ES6 methods. And yes, everything in this module is faster than ES5+ equivalents.

Reply:
Gotcha, makes sense for how they're used and not like they're a behemoth library so no objection to them being included. Was a little concerned that this was going to be a bit of a newbie-trap, but shouldn't be a concern.